core/src/avm2/globals/xml.rs

```diff
@@ -1,3 +1,81 @@
 //! XML builtin and prototype
 
+use crate::avm2::e4x::E4XNode;
 pub use crate::avm2::object::xml_allocator;
+use crate::avm2::object::{QNameObject, TObject};
+use crate::avm2::{Activation, Error, Object, QName, Value};
+
+pub fn init<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let this = this.unwrap().as_xml_object().unwrap();
+    let value = args[0];
+
+    match E4XNode::parse(value, activation) {
+        Ok(nodes) => {
+            if nodes.len() != 1 {
+                return Err(Error::RustError(
+                    format!(
+                        "XML constructor must be called with a single node: found {:?}",
+                        nodes
+                    )
+                    .into(),
+                ));
+            }
+            this.set_node(activation.context.gc_context, nodes[0])
+        }
+        Err(e) => {
+            return Err(Error::RustError(
+                format!("Failed to parse XML: {e:?}").into(),
+            ))
+        }
+    }
+
+    Ok(Value::Undefined)
+}
+
+pub fn name<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let node = this.unwrap().as_xml_object().unwrap();
+    if let Some(local_name) = node.local_name() {
+        // FIXME - use namespace
+        let namespace = activation.avm2().public_namespace;
+        Ok(QNameObject::from_qname(activation, QName::new(namespace, local_name))?.into())
+    } else {
+        Ok(Value::Null)
+    }
+}
+
+pub fn local_name<'gc>(
+    _activation: &mut Activation<'_, 'gc>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let node = this.unwrap().as_xml_object().unwrap();
```

Comment:
Same as in another review (I think?) - IMO let's do the `if let` dance like everywhere else and decide on a direction to refactor it all separately. At least that's my preference, but not a hard blocker if you leave it like this.